### ISSUE TITLE
funnel doesn't work when tailnet have custom DERP server

### ISSUE THREAD

**Issue by dionorgua** [tailscale/tailscale#14504]
### What is the issue?

I've custom DERP server configured in addition to standard servers (`"OmitDefaultRegions": false` in ACL).

Now I'm trying to take a look to `funnel` feature. And found that it doesn't work. ACL is configured correctly (it's enough to not do anything, tailscale client will add rules automatically).

I've enabled funnel using `tailscale funnel /tmp/foo` command. And now trying to access it on machine without tailscale client (after a few hours to make sure all DNS records are updated):

```
--2024-12-31 21:32:43--  https://machine.mytailnet.ts.net/
Certificates loaded: 146
Resolving machine.mytailnet.ts.net (machine.mytailnet.ts.net)... 208.111.34.11, 2607:f740:0:3f::2f0
Caching machine.mytailnet.ts.net => 208.111.34.11 2607:f740:0:3f::2f0
Connecting to machine.mytailnet.ts.net (machine.mytailnet.ts.net)|208.111.34.11|:443... connected.
Created socket 3.
Releasing 0x0000563fa82e6a30 (new refcount 1).
GnuTLS: The TLS connection was non-properly terminated.
Closed fd 3
Unable to establish SSL connection.
```

This was first time I was trying to use funnel, so I spend some time trying to fix ACL rules, etc. But I've found that just removing custom DERP server from ACL fixes it immediately.

Now I think it's most likely expected behavior because that virtual `funnel` node don't know about custom DERP server. ideally I think this should be handler more cleanly like:

- Refuse to enable funnel if custom DERP server is added
- Somehow ignore that DERP server for funnel-enabled nodes (if possible)

### Steps to reproduce

- Add custom DERP server to tailnet
- Enable funnel on node that is close to this DERP server
- Try to access it on machine without tailscale client

### Are there any recent changes that introduced the issue?

no

### OS

Linux

### OS version

Debian Unstable

### Tailscale version

1.78.1

### Bug report

BUG-3f82700e945a0d0af55f0fcb41976195029205f6d9eb0adf1ff72568278534eb-20241231213801Z-59ebe35d69346dd5

**Comment by maple3142** [comment]
I am having the same problem. Whenever a custom DERP is enabled in my tailnet, the service exposed by funnel can't be accessed from devices outside of the tailnet.

**Comment by BH2WFR** [comment]
same problem, when I deleted my custom derp server, funnel worked, otherwise not.

**Comment by secret08** [comment]
I also encountered this issue, and I was struggling for a long time because I was unaware of this bug. At the very least, there should be a prompt indicating a conflict with derp. Ideally, funnel should be able to pass through derp's traffic

**Comment by aquiveal** [comment]
Hi all, looking into this issue! It seems that when a custom DERP server is added to the tailnet ACL, `magicsock` might be preferring or routing through the custom DERP for connections that the Tailscale Funnel ingress nodes aren't aware of or can't reach. 

Because the virtual Funnel nodes run on Tailscale's infrastructure, they only know how to route traffic back to your node via the standard Tailscale DERP servers. If your node has shifted its "home DERP" to your custom DERP server, the Funnel node won't be able to establish the reverse connection.

To help confirm this hypothesis, could someone experiencing the issue run the following commands while the custom DERP is enabled and Funnel is failing?

1. Run `tailscale netcheck` and share the output. Specifically, I'm looking to see which region is listed as your "Nearest DERP" (to see if it selected your custom DERP as its home).
2. Run `tailscale ping <your-funnel-node>` (or check `tailscale status`) to see what paths are being attempted to the Funnel ingress nodes.

As a potential workaround to test the hypothesis: if you force your node to prefer a standard Tailscale DERP region using the debug command, does Funnel start working again even with the custom DERP still in the ACL?
You can try forcing it to a standard region (like region 1 or 2, assuming they are standard) by running:
`tailscale debug force-prefer-derp <region-id>`

**Comment by secret08** [comment]
Report:
  * Time: 2026-03-10T08:34:18.150076912Z
  * UDP: true
  * IPv4: yes, xx.xx.xx.xx:51174
  * IPv6: no, but OS has support
  * MappingVariesByDestIP: false
  * PortMapping:
  * CaptivePortal: false
  * Nearest DERP:
  * DERP latency:
    - MyDerp: 700µs   ()
    - nue: 152.4ms (Nuremberg)
    - sfo: 161.8ms (San Francisco)
    - hel: 172.4ms (Helsinki)
    - sea: 176.9ms (Seattle)
    - lax: 177.3ms (Los Angeles)

tailscale ping https://xx.tail53b653.ts.net/
2026/03/10 16:37:51 error looking up IP of "https://xx.tail53b653.ts.net/": lookup https://xx.tail53b653.ts.net/: no such host

tailscale debug derp-map
"1": {
			"RegionID": 1,
			"RegionCode": "nyc",
			"RegionName": "New York City",
			"Latitude": 40.7128,
			"Longitude": -74.006,
			"Nodes": [
				{
					"Name": "1f",
					"RegionID": 1,
					"HostName": "derp1f.tailscale.com",
					"IPv4": "199.38.181.104",
					"IPv6": "2607:f740:f::bc",
					"CanPort80": true
				}


tailscale debug force-prefer-derp 1

After enabling force-prefer-derp, access is possible. 
However, there is an additional piece of information: 
since the New York City node is not reachable from my network, 
I have additionally used a proxy. 
This is also the reason why I need to use derp, 
as most nodes are inaccessible to me

**Comment by aquiveal** [comment]
Awesome, thanks for running those tests @secret08! That pretty much confirms exactly what I suspected. 

Since forcing the node back to a standard Tailscale region fixed the connection, the issue is definitely the routing. 

The main thing to keep in mind here is that a custom DERP does not mean funnel ingress. They are totally different pieces of infrastructure and your custom server isn't essentially part of Tailscale's core DERP network. Tailscale's public Funnel nodes are the ones catching the web traffic, but they have no idea how to route that traffic back to your private custom DERP.

Glad the `force-prefer-derp` workaround gets you back online for now! Hopefully the Tailscale team can add a fix for this soon, maybe making it automatically fall back to standard DERPs for Funnel traffic instead of just breaking.

**Comment by aquiveal** [comment]
I’ve started using for this that works really well as a permanent fix.

Instead of relying on Funnel, you can get a custom domain (I have 3 and I don't like the cost) and add it to both Tailscale and Cloudflare with the exact same configuration.

Basically, you set it up so that when you're connected to your tailnet, your traffic routes directly and securely through Tailscale. But when you are off the tailnet (or sharing it with someone publicly), the traffic goes through Cloudflare's public proxy.

The best part is that you get to use the exact same domain name for both. The routing handles itself based on whether your Tailscale client is connected or not, so you never have to switch URLs around depending on where you are.

**Comment by dionorgua** [comment]
@aquiveal can you share more details about custom domain being added to tailscale and cloudflare?  

As far as I know there is no way to add custom domain as tailnet.. Surely it's possible to selfhost forward proxy somewhere. But for me showstopper here is lack of easy way to easily obtain TLS certificates.. 

Even if I agree to terminate TLS on such gateway it'll not be fully transparent like accessing same resource over tailnet and over internet.. Except if you agree to always do roundtrip to gateway.

**Comment by aquiveal** [comment]
@dionorgua

Yes, sadly if you want it to work seamlessly on both the public internet and Tailscale using the same domain, you do have to set up a reverse proxy (like Nginx Proxy Manager or Caddy). It becomes the single point of entry, so you're right, it's not an entirely direct 1:1 connection to the service anymore—the proxy sits in the middle to handle the routing and the TLS certs.
